README.md
## pgx_json_schema

A [JSON Schema](https://json-schema.org/) validator for Postgres implemented in Rust

This repo is a lightweight connection between the following excellent packages:
* PGX framework for developing PostgreSQL extensions in Rust [(pgx crate)](https://docs.rs/pgx/)
  
  https://github.com/zombodb/pgx
* jsonschema-rs Rust schema validation library [(jsonschema crate)](https://docs.rs/jsonschema/)
  
  https://github.com/Stranger6667/jsonschema-rs 

Supported drafts:

* Draft 7 (except optional idn-hostname)
* Draft 6
* Draft 4 (except optional bignum)

Bonus support added for:
* [JSON Type Definition (JTD)](https://jsontypedef.com/) via the [jtd crate](https://docs.rs/jtd/)
* [Apache Avro](https://avro.apache.org/) via the [avro_rs crate](https://docs.rs/avro-rs/)

### Installation:

```shell
# Install Rust
# https://www.rust-lang.org/tools/install
curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh

# Install PGX
cargo install cargo-pgx
cargo pgx init

# Download this repo
curl -L 'https://github.com/jefbarn/pgx_json_schema/archive/refs/tags/0.1.0.tar.gz' \
   | tar -xz --strip-components=1
   
# Build and install the extension package
cargo pgx package

# Enable the extension in your database
create extension pgx_json_schema;
```

### How to use:

#### JSON Schema
```
select * from json_schema_is_valid('{"maxLength": 5}'::jsonb, '"foobar"'::jsonb);

json_schema_is_valid
----------------------
f

select * from json_schema_get_errors('{"maxLength": 5}'::jsonb, '"foobar"'::jsonb);

error_value |             description              |        details         | instance_path | schema_path
------------+--------------------------------------+------------------------+---------------+-------------
"foobar"    | "foobar" is longer than 5 characters | MaxLength { limit: 5 } |               | /maxLength
```

#### JSON Type Definition

> **_NOTE:_**  The jtd library only reports the position of the validation errors, not a description.

```
select jtd_is_valid('{
    "properties": {
        "name": { "type": "string" },
        "age": { "type": "uint32" },
        "phones": {
            "elements": {
                "type": "string"
            }
        }
    }
}'::jsonb, '{
    "age": "43",
    "phones": ["+44 1234567", 442345678]
}'::jsonb);

 jtd_is_valid 
--------------
 f

select instance_path, schema_path from jtd_get_errors('{
    "properties": {
        "name": { "type": "string" },
        "age": { "type": "uint32" },
        "phones": {
            "elements": {
                "type": "string"
            }
        }
    }
}', '{
    "age": "43",
    "phones": ["+44 1234567", 442345678]
}'::jsonb);

 instance_path |           schema_path            
---------------+----------------------------------
 /age          | /properties/age/type
               | /properties/name
 /phones/1     | /properties/phones/elements/type

```

#### Apache Avro

> **_NOTE:_**  The avro library only does complete validation, there is no way to list the errors.

```
select avro_is_valid('{
    "type": "record",
    "name": "test",
    "fields": [
        {"name": "a", "type": "long", "default": 42},
        {"name": "b", "type": "string"}
    ]
}'::jsonb, '{
    "a": 27,
    "b": "foo"
}'::jsonb);

 avro_is_valid 
---------------
 t
```

### Things left to do:

- [ ] Use shared memory to store compiled validator (potential performance gain)
- [ ] More testing
- [ ] Benchmarking
- [x] Add more schema types like [JTD](https://jsontypedef.com/) and [Avro](https://avro.apache.org/)
- [ ] Support newer JSON Schema drafts
- [x] Add Dockerfile with installation example

### Prior Art
- https://github.com/gavinwahl/postgres-json-schema
- https://github.com/furstenheim/is_jsonb_valid

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.2.0] - 2021-07-28
### Added
- JSON Type Definition and Apache Avro support

## [0.1.0] - 2021-07-27
### Added
- Initial testing release

src/avro.rs
use avro_rs::types::Record;
use avro_rs::{Schema, Writer};
use pgx::*;

#[pg_extern]
fn avro_is_valid(schema: JsonB, instance: JsonB) -> bool {
    let parsed_schema =
        Schema::parse(&schema.0).unwrap_or_else(|e| panic!("Error parsing schema: {:?}", e));
    let mut writer = Writer::new(&parsed_schema, Vec::new());

    let val;
    if schema.0["type"] == "record" {
        let mut record = Record::new(&parsed_schema).unwrap();
        for (k, v) in instance.0.as_object().unwrap() {
            record.put(k, avro_rs::types::Value::from(v.clone()))
        }
        val = avro_rs::types::Value::from(record)
    } else {
        val = avro_rs::types::Value::from(instance.0);
    }

    writer.append(val).is_ok()
}

#[cfg(any(test, feature = "pg_test"))]
mod tests {
    use pgx::*;
    use std::panic;

    #[pg_test]
    fn test_avro_is_valid_bad_schema() {
        let result = panic::catch_unwind(|| {
            let _valid = Spi::get_one::<bool>(
                r#"
                select avro_is_valid('{
                    "type": "record",
                    "name": "test",
                    "fields": [
                        {"name": "a", "type": "number", "default": 42},
                        {"name": "b", "type": "string"}
                    ]
                }'::jsonb, '{
                    "a": 42,
                    "b": "foo"
                }'::jsonb)"#,
            );
        });
        assert!(result.is_err());
    }

    #[pg_test]
    fn test_avro_is_valid_true() {
        let valid = Spi::get_one::<bool>(
            r#"
            select avro_is_valid('{
                "type": "record",
                "name": "test",
                "fields": [
                    {"name": "a", "type": "long", "default": 42},
                    {"name": "b", "type": "string"}
                ]
            }'::jsonb, '{
                "a": 27,
                "b": "foo"
            }'::jsonb)"#,
        );
        assert_eq!(valid, Some(true))
    }

    #[pg_test]
    fn test_avro_is_valid_false() {
        let valid = Spi::get_one::<bool>(
            r#"
            select avro_is_valid('{
                "type": "record",
                "name": "test",
                "fields": [
                    {"name": "a", "type": "long", "default": 42},
                    {"name": "b", "type": "string"}
                ]
            }'::jsonb, '{
                "a": "27",
                "b": "foo"
            }'::jsonb)"#,
        );
        assert_eq!(valid, Some(false))
    }

    // #[pg_test]
    // fn test_avro_get_errors() {
    //     let (_value, description) = Spi::get_two::<JsonB, String>(
    //         "select * from avro_get_errors('{\"maxLength\": 5}', '\"foobar\"'::jsonb)",
    //     );
    //     assert_eq!(
    //         description,
    //         Some("\"foobar\" is longer than 5 characters".to_string())
    //     )
    // }
}


src/json_schema.rs
use pgx::*;
use jsonschema::JSONSchema;

#[pg_extern]
fn json_schema_is_valid(schema: JsonB, instance: JsonB) -> bool {
    jsonschema::is_valid(&schema.0, &instance.0)
}

#[pg_extern]
fn json_schema_get_errors(
    schema: JsonB,
    instance: JsonB,
) -> impl std::iter::Iterator<
    Item = (
        name!(error_value, JsonB),
        name!(description, String),
        name!(details, String),
        name!(instance_path, String),
        name!(schema_path, String),
    ),
> {
    let parsed_schema = JSONSchema::compile(&schema.0)
        .unwrap_or_else(|err| panic!("Error compiling schema: {:#?}", err));

    let result = parsed_schema
        .validate(&instance.0)
        .err()
        .into_iter()
        .flat_map(|iter| iter);

    let new: Vec<(JsonB, String, String, String, String)> = result
        .map(|e| {
            let description = e.to_string();
            (
                JsonB(e.instance.into_owned()),
                description,
                format!("{:?}", (e.kind)).clone(),
                e.instance_path.to_string(),
                e.schema_path.to_string(),
            )
        })
        .collect();

    new.into_iter()
}

#[cfg(any(test, feature = "pg_test"))]
mod tests {
    use pgx::*;

    #[pg_test]
    fn test_json_schema_is_valid() {
        let valid = Spi::get_one::<bool>(
            "select json_schema_is_valid('{\"maxLength\": 5}', '\"foobar\"'::jsonb)",
        );
        assert_eq!(valid, Some(false))
    }

    #[pg_test]
    fn test_json_schema_get_errors() {
        let (_value, description) = Spi::get_two::<JsonB, String>(
            "select * from json_schema_get_errors('{\"maxLength\": 5}', '\"foobar\"'::jsonb)",
        );
        assert_eq!(description, Some("\"foobar\" is longer than 5 characters".to_string()))
    }
}


src/json_type_def.rs
use jtd::Schema;
use pgx::*;

#[pg_extern]
fn jtd_is_valid(schema: JsonB, instance: JsonB) -> bool {
    let parsed_schema =
        Schema::from_serde_schema(serde_json::from_value(schema.0).unwrap()).unwrap();
    let result = jtd::validate(&parsed_schema, &instance.0, Default::default()).unwrap();

    result.is_empty()
}

#[pg_extern]
fn jtd_get_errors(
    schema: JsonB,
    instance: JsonB,
) -> impl std::iter::Iterator<Item = (name!(instance_path, String), name!(schema_path, String))> {
    let parsed_schema =
        Schema::from_serde_schema(serde_json::from_value(schema.0).unwrap()).unwrap();
    let result = jtd::validate(&parsed_schema, &instance.0, Default::default()).unwrap();

    let new: Vec<(String, String)> = result
        .into_iter()
        .map(|e| {
            let (instance_path, schema_path) = e.into_owned_paths();
            (
                if instance_path.is_empty() {
                    String::new()
                } else {
                    "/".to_owned() + &instance_path.join("/")
                },
                if schema_path.is_empty() {
                    String::new()
                } else {
                    "/".to_owned() + &schema_path.join("/")
                },
            )
        })
        .collect();

    new.into_iter()
}

#[cfg(any(test, feature = "pg_test"))]
mod tests {
    use pgx::*;

    #[pg_test]
    fn test_jtd_is_valid() {
        let valid = Spi::get_one::<bool>(
            r#"
            select jtd_is_valid('{
                "properties": {
                    "name": { "type": "string" },
                    "age": { "type": "uint32" },
                    "phones": {
                        "elements": {
                            "type": "string"
                        }
                    }
                }
            }'::jsonb, '{
                "age": "43",
                "phones": ["+44 1234567", 442345678]
            }'::jsonb)"#,
        );
        assert_eq!(valid, Some(false))
    }

    #[pg_test]
    fn test_jtd_get_errors() {
        let (instance_path, schema_path) = Spi::get_two::<String, String>(
            r#"
            select instance_path, schema_path from jtd_get_errors('{
                "properties": {
                    "name": { "type": "string" },
                    "age": { "type": "uint32" },
                    "phones": {
                        "elements": {
                            "type": "string"
                        }
                    }
                }
            }', '{
                "age": "43",
                "phones": ["+44 1234567", 442345678]
            }'::jsonb)"#,
        );

        assert_eq!(instance_path, Some("/age".to_string()));
        assert_eq!(schema_path, Some("/properties/age/type".to_string()));
    }
}
